lib/trie/trie_test.go

```diff
@@ -62,46 +79,734 @@ const (
+				&node.Leaf{ // leaf at child index 7
+					Key:   []byte{0xd},
+					Value: []byte("top leaf"),
+				},
+				nil,
+				&node.Branch{ // branch without value at child index 9
+					Key: []byte{0xe},
+					Children: [16]node.Node{
+						nil, nil, nil,
+					},
+				},
+			},
+		}
+
+		trie := NewTrie(root)
+
+		entries := trie.Entries()
+
+		expectedEntries := map[string][]byte{
+			string([]byte{0xab}):             []byte("root"),
+			string([]byte{0xab, 0x3b}):       []byte("branch 1"),
+			string([]byte{0xab, 0x7d}):       []byte("top leaf"),
+			string([]byte{0xab, 0x3b, 0x3c}): []byte("bottom leaf"),
+		}
+
+		entriesMatch(t, expectedEntries, entries)
+	})
+
+	t.Run("end to end", func(t *testing.T) {
+		t.Parallel()
+
+		trie := &Trie{
+			root:        nil,
+			childTries:  make(map[common.Hash]*Trie),
+			deletedKeys: make([]common.Hash, 0),
+		}
+
+		kv := map[string][]byte{
+			"ab":  []byte("pen"),
+			"abc": []byte("penguin"),
+			"hy":  []byte("feather"),
+			"cb":  []byte("noot"),
+		}
+
+		for k, v := range kv {
+			trie.Put([]byte(k), v)
+		}
+
+		entries := trie.Entries()
+
+		assert.Equal(t, kv, entries)
+	})
+}
+
+func Test_NextKey(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		trie    *Trie
+		key     []byte
+		nextKey []byte
+	}{
+		"nil root and nil key returns nil": {
+			trie: &Trie{},
+		},
+		"nil root returns nil": {
+			key:  []byte{1},
+			trie: &Trie{},
+		},
+		"nil key returns root leaf": {
+			trie: &Trie{
+				root: &node.Leaf{
+					Key: []byte{1},
+				},
+			},
+			nextKey: []byte{1}, // TODO should this be nil?
+		},
+		"nil key returns root branch": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{1},
+				},
+			},
+			// nextKey: []byte{1}, TODO it returns nil for now
+		},
+		"key greater than root branch full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x01},
+				},
+			},
+			key: []byte{0x02},
+		},
+		"key equal to root branch full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x01},
+				},
+			},
+			key: []byte{0x01},
+		},
+		"key smaller than root branch full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x02},
+				},
+			},
+			key: []byte{0x01},
+			// nextKey: []byte{0x02}, TODO it returns nil for now
+		},
+		"key smaller than leaf full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x01},
+					Children: [16]node.Node{
+						nil, nil,
+						&node.Leaf{
+							// full key: 0x012a
+							Key: []byte{0xa},
+						},
+					},
+				},
+			},
+			key:     []byte{0x01, 0x20},
+			nextKey: []byte{0x01, 0x2a},
+		},
+		"key equal to leaf full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x01},
+					Children: [16]node.Node{
+						nil, nil,
+						&node.Leaf{
+							// full key: 0x012a
+							Key: []byte{0xa},
+						},
+					},
+				},
+			},
+			key:     []byte{0x01, 0x2a},
+			nextKey: []byte{0x01, 0x2a}, // TODO should this return nil?
```

Comment:
I think same issue here with the nibbles - node key is in nibble form

Reply:
Hmm not sure here.

The input is `[0x01, 0x2a]` :arrow_right: (nibbles) `[0x01, 0x02, 0x0a]` which should hit the leaf which has full key in nibbles `[0x01, 0x02, 0x0a]` right?

Reply:
the input `[0x01, 0x2a]` would become `[0x00, 0x01, 0x02, 0x0a]` in nibbles I believe (correct me if I'm wrong), so the next key is correct here

Reply:
Ah yes thanks.

I needed key `[0x12, 0xa0]` which converts to `[1, 2, a, 0]` in nibbles for it to work. I really had a hard time getting I had to add trailing `0` :smile: